administration/src/cards/CardBlueprint.ts

```diff
@@ -5,144 +5,205 @@ import { CardExtensions, CardInfo } from '../generated/card_pb'
 import { Region } from '../generated/graphql'
 import { CardConfig } from '../project-configs/getProjectConfig'
 import PlainDate from '../util/PlainDate'
-import RegionExtension from './extensions/RegionExtension'
-import StartDayExtension from './extensions/StartDayExtension'
-import { Extension, ExtensionInstance, JSONExtension, findExtension } from './extensions/extensions'
+import { REGION_EXTENSION_NAME } from './extensions/RegionExtension'
+import Extensions, { Extension, ExtensionKey, ExtensionState, InferExtensionStateType } from './extensions/extensions'
 
 // Due to limited space on the cards
 const MAX_NAME_LENGTH = 30
 // Due to limited space on the qr code
 const MAX_ENCODED_NAME_LENGTH = 50
 
-export type JSONCardBlueprint<E = ExtensionInstance> = {
+export type CardBlueprint = {
   id: number
   fullName: string
-  expirationDate: string | null
-  extensions: (E extends Extension<infer T, unknown> ? JSONExtension<T> : never)[]
+  // TODO null necessary? csv?
+  expirationDate: number | null
+  extensions: Partial<ExtensionState>
 }
 
-/**
- * Blueprint for a new card. This object contains data about a future card, which will be created.
- */
-export class CardBlueprint {
-  id: number
-  fullName: string
-  expirationDate: PlainDate | null
-  extensions: ExtensionInstance[]
-  cardConfig: CardConfig
-
-  constructor(fullName: string, cardConfig: CardConfig, initParams?: Parameters<CardBlueprint['initialize']>) {
-    this.cardConfig = cardConfig
-    this.fullName = fullName
-    this.expirationDate = initParams ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity) : null
-    this.extensions = cardConfig.extensions.map(Extension => new Extension())
-    this.id = Math.floor(Math.random() * 1000000) // Assign some random ID
-    if (initParams) {
-      this.initialize(...initParams)
-    }
+type CardBlueprintKey = keyof Omit<Omit<CardBlueprint, 'extensions'>, 'id'>
+type KeyType = CardBlueprintKey | ExtensionKey | null
+
+const createRandomId = () => Math.floor(Math.random() * 1000000)
+
+export const initializeCardBlueprint = (
+  cardConfig: CardConfig,
+  region: Region | undefined = undefined,
+  { id, fullName, expirationDate, extensions }: Partial<CardBlueprint> = {}
+): CardBlueprint => {
+  // TODO Why only if region is defined?
+  // const defaultExpirationDate = region ? PlainDate.fromLocalDate(new Date()).add(cardConfig.defaultValidity).toDaysSinceEpoch() : null
```

Comment:
Why are we setting a expiration date only if a region is passed?

Reply:
tbh i have no idea. i was also really confused about this initialization stuff.
We have to pass this region in general because it will not be provided by form input nor csv file. And we get it from login information

Reply:
Okay, I'll change this to always initialize it.